File: scripts/orbit_audit.py
from __future__ import annotations

import json
import logging
from datetime import datetime, timezone
from typing import Any


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def build_audit_record(
    *,
    service: str,
    event: str,
    request_id: str,
    method: str,
    path: str,
    status_code: int,
    duration_ms: int,
    org_id: str | None = None,
    workspace_id: str | None = None,
    user_id: str | None = None,
    extra: dict[str, Any] | None = None,
) -> dict[str, Any]:
    record = {
        "timestamp": _now_iso(),
        "record_type": "audit_log",
        "service": service,
        "event": event,
        "request_id": request_id,
        "method": method,
        "path": path,
        "status_code": int(status_code),
        "duration_ms": int(duration_ms),
        "org_id": org_id,
        "workspace_id": workspace_id,
        "user_id": user_id,
    }
    if extra:
        record["extra"] = extra
    return record


def get_audit_logger(name: str) -> logging.Logger:
    logger = logging.getLogger(name)
    logger.setLevel(logging.INFO)
    logger.propagate = False
    if not logger.handlers:
        handler = logging.StreamHandler()
        handler.setFormatter(logging.Formatter("%(message)s"))
        logger.addHandler(handler)
    return logger


def emit_audit(logger: logging.Logger, record: dict[str, Any]) -> None:
    logger.info(json.dumps(record, default=str))

File: scripts/orbit-audit.js
const { randomUUID } = require('node:crypto');

function isoNow() {
  return new Date().toISOString();
}

function resolveRequestId(req) {
  return req.get('X-Request-Id') || randomUUID();
}

function buildAuditRecord({
  service,
  event,
  requestId,
  method,
  path,
  statusCode,
  durationMs,
  orgId = null,
  workspaceId = null,
  userId = null,
  extra = null,
}) {
  const record = {
    timestamp: isoNow(),
    record_type: 'audit_log',
    service,
    event,
    request_id: requestId,
    method,
    path,
    status_code: Number(statusCode || 0),
    duration_ms: Number(durationMs || 0),
    org_id: orgId,
    workspace_id: workspaceId,
    user_id: userId,
  };
  if (extra) {
    record.extra = extra;
  }
  return record;
}

function emitAudit(record) {
  console.log(JSON.stringify(record));
}

module.exports = {
  resolveRequestId,
  buildAuditRecord,
  emitAudit,
};
